Libraries/Opc.Ua.Security.Certificates/Common/AsnUtils.cs

```diff
@@ -0,0 +1,160 @@
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * The complete license agreement can be found here:
+ * http://opcfoundation.org/License/MIT/1.00/
+ * ======================================================================*/
+
+using System;
+using System.Formats.Asn1;
+using System.Text;
+
+namespace Opc.Ua.Security.Certificates
+{
+    internal static class AsnUtils
+    {
+        /// <summary>
+        /// Converts a buffer to a hexadecimal string.
+        /// </summary>
+        internal static string ToHexString(this byte[] buffer, bool bigEndian = false)
+        {
+            if (buffer == null || buffer.Length == 0)
+            {
+                return String.Empty;
+            }
+
+            StringBuilder builder = new StringBuilder(buffer.Length * 2);
+
+            if (bigEndian)
+            {
+                for (int ii = buffer.Length - 1; ii >= 0; ii--)
+                {
+                    builder.AppendFormat("{0:X2}", buffer[ii]);
+                }
+            }
+            else
+            {
+                for (int ii = 0; ii < buffer.Length; ii++)
+                {
+                    builder.AppendFormat("{0:X2}", buffer[ii]);
+                }
+            }
+
+            return builder.ToString();
+        }
+
+        /// <summary>
+        /// Converts a hexadecimal string to an array of bytes.
+        /// </summary>
+        internal static byte[] FromHexString(this string buffer)
+        {
+            if (buffer == null)
+            {
+                return null;
+            }
+
+            if (buffer.Length == 0)
+            {
+                return new byte[0];
+            }
+
+            string text = buffer.ToUpperInvariant();
+            const string digits = "0123456789ABCDEF";
+
+            byte[] bytes = new byte[(buffer.Length / 2) + (buffer.Length % 2)];
+
+            int ii = 0;
+
+            while (ii < bytes.Length * 2)
+            {
+                int index = digits.IndexOf(buffer[ii]);
+
+                if (index == -1)
+                {
+                    break;
+                }
+
+                byte b = (byte)index;
+                b <<= 4;
+
+                if (ii < buffer.Length - 1)
+                {
+                    index = digits.IndexOf(buffer[ii + 1]);
+
+                    if (index == -1)
+                    {
+                        break;
+                    }
+
+                    b += (byte)index;
+                }
+
+                bytes[ii / 2] = b;
+                ii += 2;
+            }
+
+            return bytes;
+        }
+
+        /// <summary>
+        /// Writer for Public Key parameters.
+        /// </summary>
+        /// <param name="writer">The writer</param>
+        /// <param name="integer">The key parameter</param>
+        internal static void WriteKeyParameterInteger(this AsnWriter writer, ReadOnlySpan<byte> integer)
+        {
+            if (integer[0] == 0)
```

Comment:
Is there a link with the documentation of the format? 

integer[0] = 0;
integer[1] = 0x80;

will use newStart = 0 (mean the whole integer), so why check `integer[0] equal zero` in the first place?

Reply:
https://www.itu.int/rec/T-REC-X.690-201508-I/en section 8.3 (Encoding of an integer value).

It's an error for the first byte to have all bits the same value if there's a second byte and they match the high bit on that byte.  (All redundant bytes must be removed, but 0080 is +128, 80 is -127)

Reply:
added link